### DISCUSSION TITLE
Did 3.x intentionally remove the ability to add columns not backed by model attributes?

### DISCUSSION THREAD

**Discussion by pocketnest-erik** [filamentphp/filament#10199] [Help] [answered]
In version 2.x, you could define a BadgeColumn in your table that was not associated with a model attribute, such as when you wanted a button that would copy text to the clipboard. The state of the column could be set by the `formatStateUsing()` function.
```
BadgeColumn::make('not_an_attribute')
->formatStateUsing(fn () => "Copy Something!")
->copyable()
->copyableState(fn () => "text i actually want copied")
```
However, after updating to version 3.x, we are required to use the `badge()` function on a `TextColumn` to achieve the same effect. Badges work fine on columns backed by model attributes, however in my use case nothing renders anymore.
```
TextColumn::make('not_an_attribute')
->badge()
->formatStateUsing(fn () => "Copy Something!")
->copyable()
->copyableState(fn () => "text i actually want copied")
```

Is this an intentional omission in 3.x? I know I can get around this with the Eloquent `boot()` method, but I didnt need to in 2.x.

**Comment by leandrocfe** [comment, marked as answer]
Use `->state()`
https://filamentphp.com/docs/3.x/tables/columns/getting-started#calculated-state

**Comment by pocketnest-erik** [comment]
Thanks a ton! That got me exactly what I needed
